aws/data_source_aws_route53_resolver_endpoint_test.go

```diff
@@ -0,0 +1,212 @@
+  enable_dns_support   = true
+  enable_dns_hostnames = true
+
+  tags = {
+    Name = "terraform-testacc-r53-resolver-vpc-%d"
+  }
+}
+
+resource "aws_subnet" "sn1" {
+  vpc_id            = aws_vpc.foo.id
+  cidr_block        = cidrsubnet(aws_vpc.foo.cidr_block, 2, 0)
+  availability_zone = data.aws_availability_zones.available.names[0]
+
+  tags = {
+      Name = "tf-acc-r53-resolver-sn1-%d"
+  }
+}
+
+resource "aws_subnet" "sn2" {
+  vpc_id            = aws_vpc.foo.id
+  cidr_block        = cidrsubnet(aws_vpc.foo.cidr_block, 2, 1)
+  availability_zone = data.aws_availability_zones.available.names[1]
+
+  tags = {
+    Name = "tf-acc-r53-resolver-sn2-%d"
+  }
+}
+
+resource "aws_subnet" "sn3" {
+  vpc_id            = aws_vpc.foo.id
+  cidr_block        = cidrsubnet(aws_vpc.foo.cidr_block, 2, 2)
+  availability_zone = data.aws_availability_zones.available.names[2]
+
+  tags = {
+      Name = "tf-acc-r53-resolver-sn3-%d"
+  }
+}
+
+resource "aws_security_group" "sg1" {
+  vpc_id = aws_vpc.foo.id
+  name   = "tf-acc-r53-resolver-sg1-%d"
+
+  tags = {
+       Name = "tf-acc-r53-resolver-sg1-%d"
+  }
+}
+
+resource "aws_security_group" "sg2" {
+  vpc_id = aws_vpc.foo.id
+  name   = "tf-acc-r53-resolver-sg2-%d"
+
+  tags = {
+      Name = "tf-acc-r53-resolver-sg2-%d"
+  }
+}
+`, rInt, rInt, rInt, rInt, rInt, rInt, rInt, rInt)
+}
+
+func testAccDataSourceRoute53ResolverEndpointConfig_initial(rInt int, direction, name string) string {
+	return composeConfig(testAccDataSourceRoute53ResolverEndpointConfig_base(rInt), fmt.Sprintf(`
+resource "aws_route53_resolver_endpoint" "foo" {
+  direction = "%s"
+  name      = "%s"
+
+  security_group_ids = [
+    aws_security_group.sg1.id,
+    aws_security_group.sg2.id,
+  ]
+
+  ip_address {
+    subnet_id = aws_subnet.sn1.id
+  }
+
+  ip_address {
+    subnet_id = aws_subnet.sn2.id
+    ip        = cidrhost(aws_subnet.sn2.cidr_block, 8)
+  }
+
+  tags = {
+    Environment = "production"
+    Usage       = "original"
+  }
+}
+
+data "aws_route53_resolver_endpoint" "foo" {
+    resolver_endpoint_id = aws_route53_resolver_endpoint.foo.id
+}
+`, direction, name))
+}
+
+func testAccDataSourceRoute53ResolverEndpointConfig_filter(rInt int, direction, name string) string {
+	return composeConfig(testAccDataSourceRoute53ResolverEndpointConfig_base(rInt), fmt.Sprintf(`
+resource "aws_route53_resolver_endpoint" "foo" {
+  direction = "%s"
+  name      = "%s"
+
+  security_group_ids = [
+    aws_security_group.sg1.id,
+    aws_security_group.sg2.id,
+  ]
+
+  ip_address {
+    subnet_id = aws_subnet.sn1.id
+  }
+
+  ip_address {
+    subnet_id = aws_subnet.sn2.id
+    ip        = cidrhost(aws_subnet.sn2.cidr_block, 8)
+  }
+
+  tags = {
+    Environment = "production"
+    Usage       = "original"
+  }
+}
+
+data "aws_route53_resolver_endpoint" "foo" {
+    filter {
+           name   = "Name"
+           values = [aws_route53_resolver_endpoint.foo.name]
+    }
+
+   depends_on = [aws_route53_resolver_endpoint.foo]
```

Comment:
2 spaces, not 3.

```suggestion
  depends_on = [aws_route53_resolver_endpoint.foo]
```

Reply:
Actually, remove the `depends_on` line otherwise I get:

```
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
    data_source_aws_route53_resolver_endpoint_test.go:47: Step 2/2 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout
        
        
        An execution plan has been generated and is shown below.
        Resource actions are indicated with the following symbols:
         <= read (data resources)
        
        Terraform will perform the following actions:
        
          # data.aws_route53_resolver_endpoint.foo will be read during apply
          # (config refers to values not yet known)
         <= data "aws_route53_resolver_endpoint" "foo"  {
              + arn          = (known after apply)
              + direction    = (known after apply)
              + id           = (known after apply)
              + ip_addresses = (known after apply)
              + name         = (known after apply)
              + status       = (known after apply)
              + vpc_id       = (known after apply)
        
              + filter {
                  + name   = "Name"
                  + values = [
                      + "tf-acc-test-6750279126925066157",
                    ]
                }
            }
        
        Plan: 0 to add, 0 to change, 0 to destroy.
--- FAIL: TestAccDataSourceAwsRoute53ResolverEndpoint_Filter (243.61s)
```

Reply:
Thanks for your patients @ewbankkit, when I did this work last year I didn't have a ```depends_on``` within the filter test, but when I came back to pick this up after your initial review the filter test started to fail on the data lookup. Which I found slightly odd. The AWS api had since changed the filter values, I think initially they where ```uppercase```. However when I read the docs both case types are supported. When I added the ```depends_on``` the error went away. I assumed something as changed within terraform at this point & not the provider.   

If I remove the ```depends_on``` I get the following output.
```
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
    data_source_aws_route53_resolver_endpoint_test.go:47: Step 2/2 error: Error running pre-apply plan: 
        Error: Your query returned no results. Please change your search criteria and try again
        
        
--- FAIL: TestAccDataSourceAwsRoute53ResolverEndpoint_Filter (16.33s)
--- PASS: TestAccDataSourceAwsRoute53ResolverEndpoint_Basic (117.56s)
FAIL
FAIL    github.com/terraform-providers/terraform-provider-aws/aws       119.311s
FAIL
```

If I leave the ```depends_on``` in the test passes.

```
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverEndpoint -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== RUN   TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
=== PAUSE TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
--- PASS: TestAccDataSourceAwsRoute53ResolverEndpoint_Basic (107.22s)
--- PASS: TestAccDataSourceAwsRoute53ResolverEndpoint_Filter (239.47s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       241.261s
```

Not sure where to go with this now, I noticed the name parameter is optional on the resource so I'm not sure if that causes a potential issue with the initial plan. If I add a second filter using ```SecurityGroupIds``` I don't see the failure.

```
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccDataSourceAwsRoute53ResolverEndpoint -timeout 120m
=== RUN   TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== PAUSE TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== RUN   TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
=== PAUSE TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Basic
=== CONT  TestAccDataSourceAwsRoute53ResolverEndpoint_Filter
--- PASS: TestAccDataSourceAwsRoute53ResolverEndpoint_Basic (118.19s)
--- PASS: TestAccDataSourceAwsRoute53ResolverEndpoint_Filter (250.14s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       252.067s
```

```
data "aws_route53_resolver_endpoint" "foo" {
    filter {
      name   = "Name"
      values = [aws_route53_resolver_endpoint.foo.name]
    }

    filter {
      name = "SecurityGroupIds"
      values = [aws_security_group.sg1.id, aws_security_group.sg2.id]
    }
}

```